metricbeat/module/windows/services/services_windows.go

```diff
@@ -0,0 +1,370 @@
+					return nil, ServiceErrno(err.(syscall.Errno))
+				}
+				bufSize += bytesNeeded
+			} else {
+
+				displayNameBuffer := new(bytes.Buffer)
+				serviceNameBuffer := new(bytes.Buffer)
+
+				for i := 0; i < int(servicesReturned); i++ {
+					serviceTemp := (*EnumServiceStatusProcess)(unsafe.Pointer(&servicesBuffer[i*sizeOfEnumServiceStatusProcess]))
+
+					serviceNameOffset := uintptr(unsafe.Pointer(serviceTemp.LpServiceName)) - (uintptr)(unsafe.Pointer(&servicesBuffer[0]))
+					displayNameOffset := uintptr(unsafe.Pointer(serviceTemp.LpDisplayName)) - (uintptr)(unsafe.Pointer(&servicesBuffer[0]))
+
+					displayNameBuffer.Reset()
+					serviceNameBuffer.Reset()
+
+					if err = sys.UTF16ToUTF8Bytes(servicesBuffer[displayNameOffset:serviceNameOffset], displayNameBuffer); err != nil {
+						return nil, err
+					}
+
+					if err = sys.UTF16ToUTF8Bytes(servicesBuffer[serviceNameOffset:lastOffset], serviceNameBuffer); err != nil {
+						return nil, err
+					}
+
+					lastOffset = displayNameOffset
+
+					service := ServiceStatus{}
+
+					service.DisplayName = displayNameBuffer.String()
+					service.ServiceName = serviceNameBuffer.String()
+
+					var state string
+
+					if stat, ok := serviceStates[ServiceState(serviceTemp.ServiceStatusProcess.DwCurrentState)]; ok {
+						state = stat
+					} else {
+						state = "Can not define State"
+					}
+					service.CurrentState = state
+
+					// Get detailed information
+					var serviceBufSize uint32
+					var serviceBytesNeeded uint32
+
+					serviceHandle, err := OpenService(handle, service.ServiceName, ServiceQueryConfig)
+					if err != nil {
+						return nil, err
+					}
+
+					if err := _QueryServiceConfig(serviceHandle, nil, serviceBufSize, &serviceBytesNeeded); err != nil {
+						if ServiceErrno(err.(syscall.Errno)) != SERVICE_ERROR_INSUFFICIENT_BUFFER {
+							if err := CloseServiceHandle(serviceHandle); err != nil {
+								return nil, err
+							}
+							return nil, err
+						}
+						serviceBufSize += serviceBytesNeeded
+						buffer := make([]byte, serviceBufSize)
+
+						for {
+							if err := _QueryServiceConfig(serviceHandle, &buffer[0], serviceBufSize, &serviceBytesNeeded); err != nil {
+								if ServiceErrno(err.(syscall.Errno)) != SERVICE_ERROR_INSUFFICIENT_BUFFER {
+									if err := CloseServiceHandle(serviceHandle); err != nil {
+										return nil, err
+									}
+									return nil, err
+								}
+								serviceBufSize += serviceBytesNeeded
+							} else {
+								serviceQueryConfig := (*QueryServiceConfig)(unsafe.Pointer(&buffer[0]))
+								service.StartType = serviceStartTypes[ServiceStartType(serviceQueryConfig.DwStartType)]
+								if err := CloseServiceHandle(serviceHandle); err != nil {
+									return nil, err
+								}
+								break
+							}
+						}
+					}
+
+					//Get uptime for service
+					if ServiceState(serviceTemp.ServiceStatusProcess.DwCurrentState) != ServiceStopped {
+
+						var processCreationTime syscall.Filetime
```

Comment:
This is what the system process metricset does so you can reuses the same code. See `ProcTime.Get()` in [elastic/gosigar](https://github.com/elastic/gosigar/blob/5a9ad8e33837b4e4c776d129142fcd5ccf41f639/sigar_windows.go#L330-L353). They system module [has logic](https://github.com/elastic/beats/blob/master/metricbeat/module/system/system_windows.go#L22) for attempting to enable the `SeDebugPrivilege`. We'll need to figure out how to reuse code and only activate it once. But if the system module is used along side this metricset it should work (assuming the appropriate privs exist).

Reply:
What about to move the logic from `system_windows.go` into a helper class `...\metricebeat\helper\windows.go`?